### PR TITLE
CAMEL-12769: Combination of File consumer with charset and Split DSL …

### PR DESCRIPTION
**Pull request by tadayosi**
…with XPath doesn't parse XML correctly

https://issues.apache.org/jira/browse/CAMEL-12769

### CODE REVIEW

**Comment by oscerd** [comment]
LGTM. I'd like to have feedback from @davsclaus and @onderson too

**Comment by davsclaus** [comment]
Have you run all the unit tests of camel-core, camel-spring and also since this is XML, then try camel-saxon and maybe camel-cxf.

File: camel-core/src/main/java/org/apache/camel/converter/IOConverter.java [hunk 1]
```diff
@@ -500,6 +478,53 @@ public static String getCharsetName(Exchange exchange) {
...
+        private final BufferedReader reader;
+        private final Charset defaultStreamCharset;
+
+        private ByteBuffer bufferBytes;
+        private CharBuffer bufferedChars = CharBuffer.allocate(4096);
+
+        public EncodingInputStream(File file, String charset) throws IOException {
+            this.file = file;
+            reader = toReader(file, charset);
+            defaultStreamCharset = defaultCharset.get();
```
[the remark this answers is not in the document]
Comment (tadayosi) [reply camel-core/src/main/java/org/apache/camel/converter/IOConverter.java]: If I understand your question correctly, I don't think it resolves the bug. We read files with encodings other than default and IIUC the purpose of `IOConverter.toInputStream(File, String)` is to convert a File with a charset to an InputStream with the JVM default charset encoding. So we should handle cases where the given charset and the default one differ.

**Comment by tadayosi** [comment]
I ran camel-core, camel-spring, camel-saxon, and camel-cxf, and all the unit tests passed. Should I run the full test suite locally?

File: camel-core/src/main/java/org/apache/camel/converter/IOConverter.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (onderson) [reply camel-core/src/main/java/org/apache/camel/converter/IOConverter.java]: yes, that's my point too. I am unsure if it is right approach. https://github.com/apache/camel/pull/2505/files#diff-cf527fa70918461b67c399c1a1be9f4eR508 encodes with the default charset even if you have BufferedReader with a given charset. or could it be the otherway around?

[the remark this answers is not in the document]
Comment (tadayosi) [reply camel-core/src/main/java/org/apache/camel/converter/IOConverter.java]: I didn't create this but the line came from the existing code. I just made the anonymous class in the method to a named static class. It was brought from https://issues.apache.org/jira/browse/CAMEL-8346 and https://issues.apache.org/jira/browse/CAMEL-8356 so we need to understand their backgrounds to see if it's really a right approach or not.

My understanding is that we already have the other method  `IOConverter.toInputStream(File)` so unless the following is the specification of the method, there is no reason for the existence of `IOConverter.toInputStream(File, String)`:
```
    /**
     * Converts the given {@link File} with the given charset to {@link InputStream} with the JVM default charset
     *
     * @param file the file to be converted
     * @param charset the charset the file is read with
     * @return the input stream with the JVM default charset
     */
```
Let's think that you have a file with `iso-8859-1` and want the input stream with `iso-8859-1` encoding. Then you can just use `IOConverter.toInputStream(File)`. `IOConverter.toInputStream(File, String)` is necessary only when you want the input stream with something other than `iso-8859-1`, but what encoding should the method return?  The only sensible choice would be `UTF-8` or the JVM default, but AFAIK `UTF-8` has a problem at CAMEL-8346.

**Comment by tadayosi** [comment]
I finally ran through `mvn clean install` on the project but got a few test failures on some components:
```
[INFO] Camel :: ElasticSearch5 (deprecated) 2.23.0-SNAPSHOT FAILURE [05:10 min]
[INFO] Camel :: Netty HTTP (deprecated) ................... FAILURE [05:05 min]
[INFO] Camel :: Tika ...................................... FAILURE [  7.036 s]
[INFO] Camel :: XChange ................................... FAILURE [  9.915 s]
[INFO] Camel :: Zookeeper ................................. FAILURE [10:04 min]
[INFO] Camel :: Zookeeper Master .......................... FAILURE [ 41.479 s]
[INFO] Camel :: Platforms :: Camel Catalog ................ FAILURE [ 20.017 s]
```
For camel-netty-http, camel-zookeeper, and camel-zookeeper-master, I re-ran their tests individually and they passed with my patch. For the others, I re-ran them without my patch and still got the same failures, so probably it's not my fault.

**Comment by davsclaus** [comment]
The CI server also have some unit test failures in zookeeper etc.
https://builds.apache.org/job/Camel/job/master/lastCompletedBuild/testReport/

LGTM

**Pull request merged**